### PR TITLE
8324775: C2 SuperWord: refactor visited sets

### PR DESCRIPTION
**Pull request by eme64**
Subtask of https://github.com/openjdk/jdk/pull/16620

Currently, the visited set is a "global" set that is reused all through SuperWord:
`_visited`, `_post_visited`, and also `_stk`.

This makes other refactorings difficult. I am refactoring all related code to make the visited sets local, with use of ResourceMark.

I am also refactoring the `independent` queries: from using recursive function calls to iterative. This is necessary, unless we want to pass the visited set around into the recursive calls (I think not!)

At the same time, I wanted the code of `find_dependence` to become easier, and closer to `independent`, hence I re-defined it to `mutually_independent`, and had to slightly adapt also its usages.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324775](https://bugs.openjdk.org/browse/JDK-8324775): C2 SuperWord: refactor visited sets (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [59a9205e](https://git.openjdk.org/jdk/pull/17594/files/59a9205e32769bd64d058dfaae4827eaa354679a)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [220fc173](https://git.openjdk.org/jdk/pull/17594/files/220fc17307d071966e4712345273b971e97823eb)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17594/head:pull/17594` \
`$ git checkout pull/17594`

Update a local copy of the PR: \
`$ git checkout pull/17594` \
`$ git pull https://git.openjdk.org/jdk.git pull/17594/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17594`

View PR using the GUI difftool: \
`$ git pr show -t 17594`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17594.diff">https://git.openjdk.org/jdk/pull/17594.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17594#issuecomment-1912400043)

### CODE REVIEW

File: src/hotspot/share/opto/superword.cpp [hunk 1]
```diff
@@ -1077,54 +1074,71 @@ bool SuperWord::isomorphic(Node* s1, Node* s2) {
...
+        worklist.push(pred);
+      }
+    }
+  }
+  return true; // not found -> independent
 }
 
-//------------------------------find_dependence---------------------
-// Is any s1 in p dependent on any s2 in p? Yes: return such a s2. No: return nullptr.
+// Are all nodes in nodes mutually independent?
```
Comment (vnkozlov) [inline src/hotspot/share/opto/superword.cpp:1110]: // Are all nodes in nodes list

**Review by vnkozlov — COMMENTED** [review]
Good refactoring. Thank you for replacing recursion in `independent_path()`.

**Review by chhagedorn — APPROVED** [review]
Looks good!

File: src/hotspot/share/opto/superword.hpp [hunk 2]
```diff
@@ -460,15 +449,13 @@ class SuperWord : public ResourceObj {
   bool isomorphic(Node* s1, Node* s2);
   // Is there no data path from s1 to s2 or s2 to s1?
   bool independent(Node* s1, Node* s2);
-  // Is any s1 in p dependent on any s2 in p? Yes: return such a s2. No: return nullptr.
-  Node* find_dependence(Node_List* p);
+  // Are all nodes in nodes mutually independent?
```
Comment (eme64) [inline src/hotspot/share/opto/superword.hpp:452]:
```suggestion
  // Are all nodes in nodes list mutually independent?
```

File: src/hotspot/share/opto/superword.cpp (hunk 1 above)
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:1110]:
```suggestion
// Are all nodes in nodes list mutually independent?
```

Comment (eme64) [reply src/hotspot/share/opto/superword.cpp]: fixed!

**Review by vnkozlov — APPROVED** [review]
Good.

**Pull request closed without merging**